crates/bevy_pbr/src/render/pbr.wgsl

```diff
@@ -158,45 +155,48 @@ fn fragment(
         );
 #endif
 
+        // V (view vector)
         pbr_input.V = V;
-        pbr_input.occlusion = occlusion;
 
-        pbr_input.flags = mesh[in.instance_index].flags;
-#ifdef DEFERRED_PREPASS
-        pbr_functions::alpha_discard(pbr_bindings::material, output_color);
-        out.deferred = pbr_deferred_functions::deferred_gbuffer_from_pbr_input(pbr_input);
-        out.deferred_lighting_pass_id = pbr_bindings::material.deferred_lighting_pass_id;
-#ifdef NORMAL_PREPASS
-        out.normal = vec4(pbr_input.N * 0.5 + vec3(0.5), 1.0);
-#endif // NORMAL_PREPASS
-#else // DEFERRED_PREPASS
-        output_color = pbr_functions::pbr(pbr_input);
-#endif // DEFERRED_PREPASS
     } else { // if UNLIT_BIT != 0
-        pbr_functions::alpha_discard(pbr_bindings::material, output_color);
-#ifdef DEFERRED_PREPASS    
-        var pbr_input = pbr_types::pbr_input_new();
+#ifdef DEFERRED_PREPASS
+        // in deferred mode, we need to fill some of the pbr input data even for unlit materials
+        // to pass through the gbuffer to the deferred lighting shader        
+        pbr_input = pbr_types::pbr_input_new();
         pbr_input.flags = mesh[in.instance_index].flags;
         pbr_input.material.flags = pbr_bindings::material.flags;
-        pbr_input.material.base_color = output_color;
         pbr_input.world_position = in.world_position;
-        pbr_input.world_normal = in.world_normal;
-        pbr_input.frag_coord = in.position;
-        out.deferred = pbr_deferred_functions::deferred_gbuffer_from_pbr_input(pbr_input);
-        out.deferred_lighting_pass_id = pbr_bindings::material.deferred_lighting_pass_id;
-#ifdef NORMAL_PREPASS
-        out.normal = vec4(in.world_normal * 0.5 + vec3(0.5), 1.0);
 #endif
-#endif // DEFERRED_PREPASS
     }
 
+    // apply alpha discard
+    // -------------------
+    // note even though this is based on the unlit color, it must be done after all texture samples for uniform control flow
```

Comment:
I don't think this is really just for unlit as it also allows lit alpha mask materials to discard. I'm not sure why it wasn't previously used in both conditions, but I think that was probably a mistake.

Reply:
> it discards at the top of fn pbr() for lit materials

it used to, yes

Reply:
The discard should really be done as soon as possible otherwise there is unnecessary fragment shader work being done. We should instead calculate the uv gradient in uniform control flow and pass it into later sampling functions.

Reply:
Yes. it is the same place that it was though (plus or minus a couple of assignments). I just want to get this done so I can update the other 0.12 prs though